crates/stages/src/pipeline/mod.rs

```diff
@@ -377,7 +377,22 @@ where
                         // beginning.
                         Ok(ControlFlow::Unwind {
                             target: prev_checkpoint.unwrap_or_default().block_number,
-                            bad_block: Some(block),
+                            bad_block: block,
+                        })
+                    } else if let StageError::ExecutionError { block, error } = err {
+                        warn!(
+                            target: "sync::pipeline",
+                            stage = %stage_id,
+                            bad_block = %block.number,
+                            "Stage encountered an execution error: {error}"
+                        );
+
+                        // We unwind because of an execution error. If the unwind itself fails, we
+                        // bail entirely, otherwise we restart the execution loop from the
+                        // beginning.
+                        Ok(ControlFlow::Unwind {
+                            target: prev_checkpoint.unwrap_or_default().block_number,
+                            bad_block: block,
```

Comment:
wdyt here @rkrasiuk ?

Reply:
why do we need to unwind in this case? if the execution error occurs, i'd expect the node to fail, because it's a bug rather than a bad block

Reply:
a bad block should not reach the execution stage, because of validation checks that we perform in headers & bodies

Reply:
We don't perform every type of block validation in the headers / bodies stage. For example the execution stage checks that the receipts root is correct (`BlockExecutionError::ReceiptRootDiff`). While we might return this error due to a bug in exec / receipt root calc / etc, we can't perform this validation until the execution stage anyways. So we need to treat the errors as recoverable and unwind.

#2903 is relevant as we should probably separate out the validation-related variants of `BlockExecutionError`, so we don't unwind if we reach an actually fatal error

Reply:
in that case, can we only check the receipt root diff error? i am very wary of catching more errors than we need to